ykman/driver_ccid.py

```diff
@@ -101,61 +115,56 @@ class CCIDDriver(AbstractDriver):
-            # Failure here likely indicates YK preview, in which case we'll
-            # later read the version via capabilities. This might be wrong and
-            # should be fixed later.
-            self._version = (5, 0, 0)  # TODO:Properly detect version
-        try:
-            self._read_serial()
-        except APDUError as e:
-            logger.error('Failed to read serial number', exc_info=e)
-
-    def _read_version(self):
-        s = self.send_apdu(0, INS_SELECT, 4, 0, AID.OTP)
-        self._version = tuple(c for c in six.iterbytes(s[:3]))
 
-    def _read_serial(self):
-        serial = self.send_apdu(0, INS_YK2_REQ, SLOT_DEVICE_SERIAL, 0)
+    def read_serial(self):
+        self.select(AID.OTP)
+        serial = self.send_apdu(0, OTP_INS.YK2_REQ, SLOT.DEVICE_SERIAL, 0)
         if len(serial) == 4:
-            self._serial = struct.unpack('>I', serial)[0]
+            return struct.unpack('>I', serial)[0]
 
-    def guess_version(self):
-        return self._version, self.key_type != YUBIKEY.NEO
-
-    def read_capabilities(self):
-        try:
-            self.send_apdu(0, INS_SELECT, 4, 0, AID.MGR)
-            capa = self.send_apdu(0, INS_YK4_CAPABILITIES, 0, 0)
-            return capa
-        except APDUError as e:
-            logger.error('Failed to read capabilities', exc_info=e)
-            return b''
-
-    def probe_capabilities_support(self):
-        capa = CAPABILITY.CCID
+    def read_version(self):
+        if self.pid.get_type() == YUBIKEY.YK4:
+            try:  # Attempt to read OTP applet version which should match
+                s = self.select(AID.OTP)
+                return tuple(c for c in six.iterbytes(s[:3]))
+            except APDUError:
+                pass
+        return None
+
+    def read_config(self):
+        if self.pid.get_type() == YUBIKEY.NEO:
+            raise NotImplementedError()
+        self.select(AID.MGR)
+        return self.send_apdu(0, MGR_INS.READ_CONFIG, 0, 0)
+
+    def write_config(self, data):
+        if self.pid.get_type() == YUBIKEY.NEO:
+            raise NotImplementedError()
+        self.select(AID.MGR)
+        self.send_apdu(0, MGR_INS.WRITE_CONFIG, 0, 0, data)
+
+    def probe_capabilities(self):
+        capa = TRANSPORT.CCID
         for aid, code in KNOWN_APPLETS.items():
             try:
-                self.send_apdu(0, INS_SELECT, 4, 0, aid)
+                self.select(aid)
                 capa |= code
                 logger.debug(
                     'Found applet: aid: %s , capability: %s', aid, code)
             except APDUError:
                 logger.debug(
                     'Missing applet: aid: %s , capability: %s', aid, code)
-                pass
         return capa
 
     def send_apdu(self, cl, ins, p1, p2, data=b'', check=SW_OK):
         header = [cl, ins, p1, p2, len(data)]
         body = list(six.iterbytes(data))
         try:
+            logger.debug('SEND: %s', b2a_hex(bytearray(header + body)))
```

Comment:
Are you sure these log statements won't inadvertently log PINs and such?

Reply:
Agreed, let's drop or comment out these log statements.